### ISSUE TITLE
FFI does not automatically rethrow exceptions from functions returning Handle

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#44315]
The test below would crash VM because `f()` returns unhandled exception instead of rethrowing it on the boundary. For UX purposes we should consider automatically rethrowing unhandled exceptions in the trampoline.

```dart
import 'dart:ffi';

import 'package:expect/expect.dart';
import 'package:ffi/ffi.dart';

typedef CallbackDartType = Object Function();
typedef CallbackNativeType = Handle Function();

Object alwaysThrows() {
  throw 'exception';
}

void alwaysCatches(CallbackDartType f) {
  try {
    print(f());
  } catch (e) {
    Expect.equals('exception', e);
    return;
  }
  Expect.isTrue(false);
}

void main() {
  final ptr = Pointer.fromFunction<CallbackNativeType>(alwaysThrows);
  final f = ptr.asFunction<CallbackDartType>();
  alwaysCatches(f);
}
```

/cc @dcharkes

**Comment by mraleph** [comment]
as @rmacnak-google points out in #45869 leaking non-Instances to Dart code in general is unsafe, so we should really fix this.

**Comment by dcharkes** [comment]
This was addressed in b113fffb65a002accf53724e5d57110ad215b5f7.
